dialog_design.md

```diff
@@ -0,0 +1,72 @@
+# Dialogs
+
+Goal: Display a message and an optional graphic to all users, or some, or one, and return a response.
+
+## Architecture
+
+Create a new `Dialog` class that manages calls and responses. It is also a good place to park static convenience
+functions and event listeners. I propose a couple different setup options:
+
+1. `text` - The text that the dialog displays.
+2. `buttons` - An array of labels for dialog buttons. When a button is clicked, this label is included in the response.
+3. `input` - Whether the dialog should present a text input. If so, the text is included in the response.
+4. `recipients` (optional) - The dialog is displayed to all currently connected users by default, but can be restricted
+	to a subset of users with this option. Takes either a `GroupMask`, a `User[]`, or just one `User`. This filtering
+	could be implemented in the sync layer fairly easily, I think.
+5. `icon` (optional) - A graphic to display alongside the text. Can be one of the built-in icons (`info`, `warning`,
+	`error`, or `question`), or a custom texture supplied by the developer.
+6. `persistent` (optional) - If `true`, the dialog is presented to users in the recipients list even if they are
+	not currently online. I imagine this would be implemented by an extra handler to `context.onUserJoined`, with
+	special logic to not send the dialog to the same user more than once. This is complicated by changing membership
+	in groups, not sure how to handle that yet.
+
+Once the dialog is set up, you would call `dialog.show()` to present it to users. If you need to hide the dialog
+programmatically, call `dialog.cancel()`.
+
+To receive the replies, the developer would listen to events on the dialog instance. I'm proposing one event fired
+when a single response comes in:
+
+```ts
+interface DialogResponse {
+	user: User;
+	button: string;
+	text?: string;
+}
+public on(eventType: 'response', listener: (r: DialogResponse) => void): this;
+```
+
+And another when all the responses are in, if it's not a persistent dialog. This version requires collecting all the
+responses inside the dialog class so they can be passed together into this callback:
+
+```ts
+public on(eventType: 'finished', listener: (responses: DialogResponse[]) => void): this;
+```
+
```

Comment:
I don't think "Finished" makes sense. The concept of "all responses" seems strange to me - you can't ever expect all users to interact with an app. If just a single user doesn't interact then this won't trigger, so this would rarely be useful

Reply:
It makes sense without the `persistent` property. You're choosing a list of users to present the dialog to, so it makes sense to expect them to be active participants. Or else you're sending it to a single user on `onUserJoined`, in which case it's interchangeable with `onResponse`.